coalib/output/printers/LogPrinter.py

```diff
@@ -1,54 +1,194 @@
+from coalib.output.printers.LOG_LEVEL import LOG_LEVEL_COLORS
 from coalib.processes.communication.LogMessage import LogMessage
 
 
+class PrinterHandler(logging.StreamHandler):
+    """
+    Handler for logging which uses printer to print messages and falls back to
+    writing to stdout if no printer is provided.
+    """
+
+    def __init__(self, printer=None, *args, **kwargs):
+        """
+        :param printer:  Printer object which will be used for printing
+                         messages.
+        """
+        self.printer = printer
+        super(PrinterHandler, self).__init__(*args, **kwargs)
+
+    def emit(self, record):
+        msg = self.format(record)
+        if self.printer:
+            self.printer.print(msg)
+        else:
+            sys.stdout.write(msg + "\n")
+
+
+class ColoredFormatter(logging.Formatter):
+    """
+    Formatter for logging which adds colors to the output by inserting special
+    symbols where appropriate.
+    """
+
+    def __init__(self, *args, **kwargs):
+        """
+        :param use_color:  If True highlights records levelname with respective
+                           color.
+        """
+        self.use_color = kwargs.pop('use_color', False)
+        super(ColoredFormatter, self).__init__(*args, **kwargs)
+
+    def format(self, record, *args, **kwargs):
+        if record.levelno in LOG_LEVEL_COLORS and self.use_color:
+            color_begin = LOG_LEVEL_COLORS[record.levelno]
+            color_end = Style.RESET_ALL
+        else:
+            color_begin = color_end = ''
+        record.levelname = '{cb}[{levelname}]{ce}'.format(
+            cb=color_begin,
+            ce=color_end,
+            levelname=record.levelname)
+        return super(ColoredFormatter, self).format(record, *args, **kwargs)
+
+
 class LogPrinter:
     """
     The LogPrinter class allows to print log messages to an underlying Printer.
 
     This class is an adapter, means you can create a LogPrinter from every
     existing Printer instance.
     """
+    _loggers = {}
+
+    @staticmethod
+    def _get_logger(printer, logger_name, time_fmt=None):
+        """
+        Returns existing or newly created and configured logger.
+        :param logger_name: name of the logger to be returned.
+        :param printer:     printer to initialize logger with if it doesn't
+                            exist, yet.
+        :param time_fmt:    format to be used for time output. No time is
+                            pinter if None.
+        """
+        if not LogPrinter._loggers.get(logger_name):
+            if time_fmt:
+                format_str = '%(levelname)s[%(asctime)s] %(message)s'
+            else:
+                format_str = '%(levelname)s %(message)s'
+
+            use_color = isinstance(printer, ColorPrinter)
+            formatter = ColoredFormatter(format_str,
+                                         time_fmt,
+                                         use_color=use_color)
```

Comment:
With the formatter we don't need the color printer anymore do we? Also we would want to avoid printing copies when piping strong I guess...